--- a/src/tap-bridge/model/tap-bridge.cc        Tue May 02 21:48:19 2017 +0200
+++ b/src/tap-bridge/model/tap-bridge.cc        Tue Sep 26 00:37:02 2017 +0200
@@ -143,7 +143,8 @@
     m_startEvent (),
     m_stopEvent (),
     m_fdReader (0),
-    m_ns3AddressRewritten (false)
+    m_ns3AddressRewritten (false),
+    m_linkUp (false)
 {
   NS_LOG_FUNCTION_NOARGS ();
   m_packetBuffer = new uint8_t[65536];
@@ -991,7 +992,7 @@
   p->CopyData (m_packetBuffer, p->GetSize ());

   uint32_t bytesWritten = write (m_sock, m_packetBuffer, p->GetSize ());
-  NS_ABORT_MSG_IF (bytesWritten != p->GetSize (), "TapBridge::ReceiveFromBridgedDevice(): Write error.");
+  NS_ABORT_MSG_IF (bytesWritten != p->GetSize () && errno != EIO, "TapBridge::ReceiveFromBridgedDevice(): Write error.");

   NS_LOG_LOGIC ("End of receive packet handling on node " << m_node->GetId ());
   return true;
